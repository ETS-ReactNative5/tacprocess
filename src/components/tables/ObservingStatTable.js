import React from "react";
import propTypes from "prop-types";

function setProposalTransparency(proposals) {
  const data = {
    maxSeeing: {
      less10: {
        timeRequests: 0,
        noOfProposals: 0,

      },
      less15: {
        timeRequests: 0,
        noOfProposals: 0,

      },
      less20: {
        timeRequests: 0,
        noOfProposals: 0,

      },
      less50: {
        timeRequests: 0,
        noOfProposals: 0,

      },
    },
    transparency: {
      clear: {
        timeRequests: 0,
        noOfProposals: 0,
        fraction: 0
      },
      thin: {
        timeRequests: 0,
        noOfProposals: 0,
        fraction: 0
      },
      thick: {
        timeRequests: 0,
        noOfProposals: 0,
        fraction: 0
      },
      any: {
        timeRequests: 0,
        noOfProposals: 0,
        fraction: 0
      }
    }
  }
  proposals.map(p => {
    let reqTime = p.totalRequestedTime
    if ( isNaN(reqTime)) { reqTime = 0 }
    if (p.transparency === "Clear"){
      data.transparency.clear.noOfProposals += 1
      data.transparency.clear.timeRequests += reqTime
    }
    if (p.transparency === "Any"){

      data.transparency.any.noOfProposals += 1
      data.transparency.any.timeRequests += reqTime
    }
    if (p.transparency === "Thick cloud"){
      data.transparency.thick.noOfProposals += 1
      data.transparency.thick.timeRequests += reqTime
    }
    if (p.transparency === "Thin cloud"){
      data.transparency.thin.noOfProposals += 1
      data.transparency.thin.timeRequests += reqTime
    }

    // seeing
    if (p.maxSeeing <= 1){
      data.maxSeeing.less10.noOfProposals += 1
      data.maxSeeing.less10.timeRequests += reqTime
    }
    if (p.maxSeeing > 1 && p.maxSeeing <= 1.5){

      data.maxSeeing.less15.noOfProposals += 1
      data.maxSeeing.less15.timeRequests += reqTime
    }
    if (p.maxSeeing > 1.5 && p.maxSeeing <= 2){
      data.maxSeeing.less20.noOfProposals += 1
      data.maxSeeing.less20.timeRequests += reqTime
    }
    if (p.maxSeeing > 2 && p.maxSeeing <= 5.0){
      data.maxSeeing.less50.noOfProposals += 1
      data.maxSeeing.less50.timeRequests += reqTime
    }


    return p
  })

  return data
}


const ObservingStatTable = (proposals) => {

  const data = setProposalTransparency(proposals.proposals)
  return(
  <div>
    <h1>Observing Statistics</h1>
    <table className="table center-table">
      <thead>
        <tr>
          <th><h2>Conditions</h2></th>
          <th><h2>Time requested (hrs)</h2></th>
          <th><h2>Number of Proposals</h2></th>
          <th><h2>Fraction of Total Request (%)</h2></th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td><h2>Clear</h2></td>
          <td><h2>{ data.transparency.clear.timeRequests }</h2></td>
          <td><h2>{ data.transparency.clear.noOfProposals }</h2></td>
          <td><h2>{0}</h2></td>
        </tr>
        <tr>
          <td><h2>Thin cloud</h2></td>
          <td><h2>{ data.transparency.thin.timeRequests }</h2></td>
          <td><h2>{ data.transparency.thin.noOfProposals }</h2></td>
          <td><h2>{0}</h2></td>
        </tr>
        <tr>
          <td><h2>Thick cloud</h2></td>
          <td><h2>{ data.transparency.thick.timeRequests }</h2></td>
          <td><h2>{ data.transparency.thick.noOfProposals }</h2></td>
          <td><h2>{0}</h2></td>
        </tr>
        <tr>
          <td><h2>Any</h2></td>
          <td><h2>{ data.transparency.any.timeRequests }</h2></td>
          <td><h2>{ data.transparency.any.noOfProposals }</h2></td>
          <td><h2>{0}</h2></td>
        </tr>
      </tbody>
      <thead>
        <tr>
          <th />
          <th />
          <th />
          <th />
        </tr>
      </thead>
      <tbody>
        <tr>
          <td><h3>Max Seeing <br /> &#x2266; 1.0 </h3></td>
          <td><h2>{ data.maxSeeing.less10.timeRequests }</h2></td>
          <td><h2>{ data.maxSeeing.less10.noOfProposals }</h2></td>
          <td><h2>{0}</h2></td>
        </tr>
        <tr>
          <td><h3>Max Seeing <br /> &#x2266; 1.5</h3></td>
          <td><h2>{ data.maxSeeing.less15.timeRequests }</h2></td>
          <td><h2>{ data.maxSeeing.less15.noOfProposals }</h2></td>
          <td><h2>{0}</h2></td>
        </tr>
        <tr>
          <td><h3>Max Seeing <br /> &#x2266; 2.0</h3></td>
          <td><h2>{ data.maxSeeing.less20.timeRequests }</h2></td>
          <td><h2>{ data.maxSeeing.less20.noOfProposals }</h2></td>
          <td><h2>{0}</h2></td>
        </tr>
        <tr>
          <td><h3>Max Seeing <br /> &#x2266; 5.0</h3></td>
          <td><h2>{ data.maxSeeing.less50.timeRequests }</h2></td>
          <td><h2>{ data.maxSeeing.less50.noOfProposals }</h2></td>
          <td><h2>{0}</h2></td>
        </tr>
      </tbody>
    </table>
  </div>
)}

ObservingStatTable.propTypes = {
  proposals: propTypes.array.isRequired
}

export default ObservingStatTable;
