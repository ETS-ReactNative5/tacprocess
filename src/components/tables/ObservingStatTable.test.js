import React from "react";
import renderer from "react-test-renderer";
import { shallow, mount, render } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import ObservingStatTable from "./ObservingStatTable";

const proposals = [
  {
    transparency: 'Clear',
    maxSeeing: 2,
    instruments: {
      scam: [ { dictatorMode: "NORMAL" } ],
    },
    timeRequests: [
      {
        semester: '2018-1',
        distribution: [ { partnerCode: 'RSA', time: 150 }, { partnerCode: 'UW', time: 200 }, { partnerCode: 'IUCAA', time: 250 } ]
      },
      {
        semester: '2018-1',
        distribution: [ { partnerCode: 'RSA', time: 400 }, { partnerCode: 'UW', time: 50 } ]
      }
    ],
    allocatedTime:
      {
        UW: { p0: 0, p1: 0, p2: 0, p3: 0, p4: 0, p0p1: 0},
        UKSC: { p0: 0, p1: 0, p2: 0, p3: 0, p4: 0, p0p1: 0}
      }
  },
  {
    transparency: 'Thin cloud',
    maxSeeing: 2,
    instruments: {
      scam: [ { dictatorMode: "NORMAL" } ],
    },
    timeRequests: [
      {
        semester: '2017-1',
        distribution: [ { partnerCode: 'RSA', time: 150 }, { partnerCode: 'UW', time: 200 }, { partnerCode: 'IUCAA', time: 250 } ]
      },
      {
        semester: '2018-1',
        distribution: [ { partnerCode: 'RSA', time: 400 }, { partnerCode: 'UW', time: 50 } ]
      }
    ]
  }
];

// Checking if the ObservingStatTable Component renders correctly for different inputs
describe("ObservingStatTable Component", () => {

  //Enzyme method testing if it renders correctly with empty value of the target
  it("Render the ObservingStatTable Component having unpopulated props with no errors", () => {
    const rendered = mount(<ObservingStatTable proposals={[]} />);
    expect(shallowToJson(rendered)).toMatchSnapshot();
  });

  it("Render the ObservingStatTable Component having populated props with no errors", () => {
    const rendered = mount(<ObservingStatTable proposals={proposals} />);
    expect(shallowToJson(rendered)).toMatchSnapshot();
  });

});
