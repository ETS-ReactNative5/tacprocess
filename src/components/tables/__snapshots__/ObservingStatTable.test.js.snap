// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObservingStatTable Component Render the ObservingStatTable Component having populated props with no errors 1`] = `
<ObservingStatTable
  proposals={
    Array [
      Object {
        "allocatedTime": Object {
          "UKSC": Object {
            "p0": 0,
            "p0p1": 0,
            "p1": 0,
            "p2": 0,
            "p3": 0,
            "p4": 0,
          },
          "UW": Object {
            "p0": 0,
            "p0p1": 0,
            "p1": 0,
            "p2": 0,
            "p3": 0,
            "p4": 0,
          },
        },
        "instruments": Object {
          "scam": Array [
            Object {
              "dictatorMode": "NORMAL",
            },
          ],
        },
        "maxSeeing": 2,
        "timeRequests": Array [
          Object {
            "distribution": Array [
              Object {
                "partnerCode": "RSA",
                "time": 150,
              },
              Object {
                "partnerCode": "UW",
                "time": 200,
              },
              Object {
                "partnerCode": "IUCAA",
                "time": 250,
              },
            ],
            "semester": "2018-1",
          },
          Object {
            "distribution": Array [
              Object {
                "partnerCode": "RSA",
                "time": 400,
              },
              Object {
                "partnerCode": "UW",
                "time": 50,
              },
            ],
            "semester": "2018-1",
          },
        ],
        "transparency": "Clear",
      },
      Object {
        "instruments": Object {
          "scam": Array [
            Object {
              "dictatorMode": "NORMAL",
            },
          ],
        },
        "maxSeeing": 2,
        "timeRequests": Array [
          Object {
            "distribution": Array [
              Object {
                "partnerCode": "RSA",
                "time": 150,
              },
              Object {
                "partnerCode": "UW",
                "time": 200,
              },
              Object {
                "partnerCode": "IUCAA",
                "time": 250,
              },
            ],
            "semester": "2017-1",
          },
          Object {
            "distribution": Array [
              Object {
                "partnerCode": "RSA",
                "time": 400,
              },
              Object {
                "partnerCode": "UW",
                "time": 50,
              },
            ],
            "semester": "2018-1",
          },
        ],
        "transparency": "Thin cloud",
      },
    ]
  }
>
  <div>
    <h1>
      Observing Statistics
    </h1>
    <table
      className="table center-table"
    >
      <thead>
        <tr>
          <th>
            <h2>
              Conditions
            </h2>
          </th>
          <th>
            <h2>
              Time requested (hrs)
            </h2>
          </th>
          <th>
            <h2>
              Number of Proposals
            </h2>
          </th>
          <th>
            <h2>
              Fraction of Total Request (%)
            </h2>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <h2>
              Clear
            </h2>
          </td>
          <td>
            <h2>
              0.00
            </h2>
          </td>
          <td>
            <h2>
              1
            </h2>
          </td>
          <td>
            <h2>
              0
            </h2>
          </td>
        </tr>
        <tr>
          <td>
            <h2>
              Thin cloud
            </h2>
          </td>
          <td>
            <h2>
              0.00
            </h2>
          </td>
          <td>
            <h2>
              1
            </h2>
          </td>
          <td>
            <h2>
              0
            </h2>
          </td>
        </tr>
        <tr>
          <td>
            <h2>
              Thick cloud
            </h2>
          </td>
          <td>
            <h2>
              0.00
            </h2>
          </td>
          <td>
            <h2>
              0
            </h2>
          </td>
          <td>
            <h2>
              0
            </h2>
          </td>
        </tr>
        <tr>
          <td>
            <h2>
              Any
            </h2>
          </td>
          <td>
            <h2>
              0.00
            </h2>
          </td>
          <td>
            <h2>
              0
            </h2>
          </td>
          <td>
            <h2>
              0
            </h2>
          </td>
        </tr>
      </tbody>
      <thead>
        <tr>
          <th />
          <th />
          <th />
          <th />
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <h3>
              Max Seeing 
              <br />
               ≦ 1.0 
            </h3>
          </td>
          <td>
            <h2>
              0.00
            </h2>
          </td>
          <td>
            <h2>
              0
            </h2>
          </td>
          <td>
            <h2>
              0
            </h2>
          </td>
        </tr>
        <tr>
          <td>
            <h3>
              Max Seeing 
              <br />
               ≦ 1.5
            </h3>
          </td>
          <td>
            <h2>
              0.00
            </h2>
          </td>
          <td>
            <h2>
              0
            </h2>
          </td>
          <td>
            <h2>
              0
            </h2>
          </td>
        </tr>
        <tr>
          <td>
            <h3>
              Max Seeing 
              <br />
               ≦ 2.0
            </h3>
          </td>
          <td>
            <h2>
              0.00
            </h2>
          </td>
          <td>
            <h2>
              2
            </h2>
          </td>
          <td>
            <h2>
              0
            </h2>
          </td>
        </tr>
        <tr>
          <td>
            <h3>
              Max Seeing 
              <br />
               ≦ 5.0
            </h3>
          </td>
          <td>
            <h2>
              0.00
            </h2>
          </td>
          <td>
            <h2>
              0
            </h2>
          </td>
          <td>
            <h2>
              0
            </h2>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</ObservingStatTable>
`;

exports[`ObservingStatTable Component Render the ObservingStatTable Component having unpopulated props with no errors 1`] = `
<ObservingStatTable
  proposals={Array []}
>
  <div>
    <h1>
      Observing Statistics
    </h1>
    <table
      className="table center-table"
    >
      <thead>
        <tr>
          <th>
            <h2>
              Conditions
            </h2>
          </th>
          <th>
            <h2>
              Time requested (hrs)
            </h2>
          </th>
          <th>
            <h2>
              Number of Proposals
            </h2>
          </th>
          <th>
            <h2>
              Fraction of Total Request (%)
            </h2>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <h2>
              Clear
            </h2>
          </td>
          <td>
            <h2>
              0.00
            </h2>
          </td>
          <td>
            <h2>
              0
            </h2>
          </td>
          <td>
            <h2>
              0
            </h2>
          </td>
        </tr>
        <tr>
          <td>
            <h2>
              Thin cloud
            </h2>
          </td>
          <td>
            <h2>
              0.00
            </h2>
          </td>
          <td>
            <h2>
              0
            </h2>
          </td>
          <td>
            <h2>
              0
            </h2>
          </td>
        </tr>
        <tr>
          <td>
            <h2>
              Thick cloud
            </h2>
          </td>
          <td>
            <h2>
              0.00
            </h2>
          </td>
          <td>
            <h2>
              0
            </h2>
          </td>
          <td>
            <h2>
              0
            </h2>
          </td>
        </tr>
        <tr>
          <td>
            <h2>
              Any
            </h2>
          </td>
          <td>
            <h2>
              0.00
            </h2>
          </td>
          <td>
            <h2>
              0
            </h2>
          </td>
          <td>
            <h2>
              0
            </h2>
          </td>
        </tr>
      </tbody>
      <thead>
        <tr>
          <th />
          <th />
          <th />
          <th />
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <h3>
              Max Seeing 
              <br />
               ≦ 1.0 
            </h3>
          </td>
          <td>
            <h2>
              0.00
            </h2>
          </td>
          <td>
            <h2>
              0
            </h2>
          </td>
          <td>
            <h2>
              0
            </h2>
          </td>
        </tr>
        <tr>
          <td>
            <h3>
              Max Seeing 
              <br />
               ≦ 1.5
            </h3>
          </td>
          <td>
            <h2>
              0.00
            </h2>
          </td>
          <td>
            <h2>
              0
            </h2>
          </td>
          <td>
            <h2>
              0
            </h2>
          </td>
        </tr>
        <tr>
          <td>
            <h3>
              Max Seeing 
              <br />
               ≦ 2.0
            </h3>
          </td>
          <td>
            <h2>
              0.00
            </h2>
          </td>
          <td>
            <h2>
              0
            </h2>
          </td>
          <td>
            <h2>
              0
            </h2>
          </td>
        </tr>
        <tr>
          <td>
            <h3>
              Max Seeing 
              <br />
               ≦ 5.0
            </h3>
          </td>
          <td>
            <h2>
              0.00
            </h2>
          </td>
          <td>
            <h2>
              0
            </h2>
          </td>
          <td>
            <h2>
              0
            </h2>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</ObservingStatTable>
`;
